### ISSUE TITLE
FFT Module (like torch.fft)

### ISSUE THREAD

**Issue by discordance** [burn-rs/burn#788]
Hello burn team!
Thank you for all the hard work!

### Feature description

It would be incredibly valuable to add would be a Discret Fourier Transform (FFT) module, similar to the torch.fft in [PyTorch](https://pytorch.org/docs/stable/fft.html).

### Feature motivation

Having it as a part of the framework would add a fundational building block when working with differentiable Digital Signal Processing (e.g dDSP), Spectral losses, and Audio related tasks.

**Comment by TheAndrewJackson** [comment]
If this feature is added this crate might be helpful: https://github.com/ejmahler/RustFFT


It has very fast SIMD accelerated FFT processing.

**Comment by antimora** [comment]
Since Burn uses different backends, we should we take advantage of this. Doing purely for CPU will have data transfer overhead for GPU backends. I believe @Gadersd has implemented something already using Burn ops. https://github.com/Gadersd/whisper-burn

Perhaps, we can implement a default for CPU and let GPU backends override with more specialized implementation.

**Comment by Gadersd** [comment]
Feel free to use my FFT implementation, but be aware that it isn’t as general as the one torch uses. There are some padding options and checks that should be added for generality.

**Comment by TomWyllie** [comment]
I've started work on a PR for this feature, based on an NdArray / Wgpu op I wrote as a custom backend a few months back. There's a few design decisions to get right. We don't support complex numbers at the moment so the way I see it there's 3 ways to handle complex inputs and outputs for each op:

1. Add a complex type as an `Element` - probably the most idiomatic but I'm not sure if this is practical
2. Require separate real and imaginary tensors
3. Require a complex dimension of length 2 as the last dimension

Secondly, I'm guessing nobody will be too upset if we don't have a backward pass for FFT? The derivative of the fourier transform of a function is itself complex, and propagating imaginary gradients around the place seems like a bad idea. I'm assuming folks will want FFT to preprocess / extract features rather than as an op that needs to be in the graph.

Thoughts?

**Comment by antimora** [comment]
Having it for the forward pass will cover the majority of use cases. As you mentioned many simply need to extract features and it is better to do using the backend device instead of CPU.

**Comment by antimora** [comment]
The only thing I could suggest is to file issues for the remaining work so others can pick up or refer to the deficiencies.

**Comment by TomWyllie** [comment]
I have a branch for an FFT op just about ready for a pull request :slightly_smiling_face:  it's implemented and working for NdArray and WGPU with a WGSL shader, I've done 1D FFT and 1D inverse FFT but will add an issue for supporting 2D soon. Haven't added candle or autodiff. There's a couple outstanding tasks before it's ready to be merged to main though:

- Rewrite the WGSL kernel into the new JIT GPU representation (@louisfd you still up for this?)
- Not sure how to register this as an op in the fusion backend. The input + output shape is the same for all the FFT ops which I suppose makes the description of the op simpler - is there an existing op that I can base the implementation on?
- I've not added support for Torch backend but we can do this down the line too, `libtorch` does support FFT.

If somebody can make an FFT branch in this repo then I can merge my fork in

**Comment by louisfd** [comment]
Hi Tom, 
I do believe fourier transform would be a nice addition to burn 😄 
I made a branch so we can work on this together: https://github.com/tracel-ai/burn/tree/feat/ops/fft

I looked at your wgsl and it won't be very hard to translate to JIT. But before that can you make sure it's thoroughly tested? The translation process is error-prone so I need tests to make sure I don't screw up.

**Comment by TomWyllie** [comment]
Sounds good. I have a bunch of tests but I'll check some test cases from other libraries tomorrow to make sure I got all edge cases, there's a lot of different parts to the algorithm.

**Comment by jbelanich** [comment]
Hello folks! This feature indeed would be pretty helpful -- I was about to experiment using burn to train some audio models and was finding myself reaching for an FFT module. I'm happy to help with such a feature. I saw there exists a [feature branch](https://github.com/tracel-ai/burn/compare/main...feat/ops/fft) but I wasn't sure what it's current state is. Are there any tasks that can be peeled off on which I can contribute?

One additional note -- it would be wonderful if autodiff functionality were included. Reading the feature branch, it seems this might be pushed off to the future. Differentiating through the fft is indeed used in many ML audio applications (spectral losses, DDSP, etc mentioned in the [original issue text](https://github.com/tracel-ai/burn/issues/788#issue-1887718338)). As a user, an implementation not including an autodiff backend would be very surprising. Again, happy to help make this a reality.

**Comment by nathanielsimard** [comment]
I think the right approach would be to provide FFT kernels in CubeCL, we can then bundle them with Burn.

**Comment by janroden** [comment]
> I think the right approach would be to provide FFT kernels in CubeCL, we can then bundle them with Burn.

That would be great to have FFT also in CubeCL!

As far as I can tell, the larger Rust ecosystem is currently lacking a portable FFT that can also run on the GPU.

For the CPU, there's [RustFFT](https://github.com/ejmahler/RustFFT) and [ndrustfft](https://github.com/preiter93/ndrustfft), a wrapper for RustFFT for complex/real n-dimensional FFTs (using ndarray).

For the GPU, we got crates like [vkfft-rs](https://github.com/semio-ai/vkfft-rs), [tch-rs](https://github.com/LaurentMazare/tch-rs/issues/313) (Rust PyTorch wrapper), and [arrayfire-rust](https://arrayfire.org/arrayfire-rust/arrayfire/fn.fft.html) a [wrapper for ArrayFire](https://github.com/arrayfire/arrayfire-rust).

But it would be great to also have a portable WebGPU FFT, to not have to install huge dependencies and be able to ship stand-alone binaries. 

FFT on GPU is, especially for large data, normally much faster than on the CPU (depending on the GPU and CPU of course). 

For instance, 3D image data in medical imaging and microscopy can be GBs and one would normally want to run FFT on the GPU then. 

Having an efficient, portable (n-dim) FFT (e.g. based on WebGPU), would also enable significantly faster convolutions on large data.

**Comment by janroden** [comment]
I just saw that you already opened [this issue](https://github.com/tracel-ai/cubecl/issues/20) (that has FFT as a to-do item) over at the CubeCL repo, so linking it here for reference.

**Comment by thedevleon** [comment]
> Hello folks! This feature indeed would be pretty helpful -- I was about to experiment using burn to train some audio models and was finding myself reaching for an FFT module. I'm happy to help with such a feature. I saw there exists a [feature branch](https://github.com/tracel-ai/burn/compare/main...feat/ops/fft) but I wasn't sure what it's current state is. Are there any tasks that can be peeled off on which I can contribute?
> 
> One additional note -- it would be wonderful if autodiff functionality were included. Reading the feature branch, it seems this might be pushed off to the future. Differentiating through the fft is indeed used in many ML audio applications (spectral losses, DDSP, etc mentioned in the [original issue text](https://github.com/tracel-ai/burn/issues/788#issue-1887718338)). As a user, an implementation not including an autodiff backend would be very surprising. Again, happy to help make this a reality.

What is the state of this branch? Seems like a lot was already implemented, will there be an PR soon? Hoping to use fft's in burn soon.

**Comment by laggui** [comment]
> What is the state of this branch? Seems like a lot was already implemented, will there be an PR soon? Hoping to use fft's in burn soon.

I wasn't aware of this branch 😅 but it looks like it was a WIP. A lot has changed since, and FFT kernels should instead be implemented in CubeCL (as tracked and discussed in https://github.com/tracel-ai/cubecl/issues/20).

The easy part after that will be to provide them with Burn 🙂

**Comment by for-lack-of-a-better-name-j** [comment]
Would be interested in a 2D inverse and forward FFT with Autodiff. Adding comment here because I want the Github notifications. I may be able to contribute to these items if I ever get any time. No promises though.

**Comment by antimora** [comment]
Closing as completed. The core FFT operations are implemented in `crates/burn-tensor/src/tensor/signal/`:
- `rfft` (1D real FFT) with cubecl kernel
- `irfft` (inverse real FFT)
- `hamming_window`
- `hann_window`

Remaining items (STFT, MelWeightMatrix, complex DFT, BlackmanWindow) are tracked in the more specific issue #4531.
